### ISSUE TITLE
pba image doesn't seem to unlock drive

### ISSUE THREAD

**Issue by jmccarv** [r0m30/msed#9]
First I want to thank you for working on this project, it looks very promising!

Hopefully I'm just missing something stupid here since I have no prior experience with OPAL / msed.

I am trying to use msed on a Dell Optiplex 7020 with a Samsung 840 EVO 250GB drive.  I am using version 0.20beta.  I followed the instructions listed on your web site and everything appeard to be working as expected.  The problem comes when I reboot the drive and the PBA image runs and prompts me for the password to unlock the drive.  Whether I put in the correct password or an incorrect one it does the same thing, it tries to chain-boot hd0,0 and then it appears to boot right back to the PBA and prompts me again for the password; almost like it's not unlocking the drive and so when it chain loads, it loads the PBA again.

I am able to follow your directions to remove OPAL (using --revertnoerase followed by reverttper) and the drive will boot as you would expect.  The same is true if I turn off locking and the PBA via --disableLockingRange 0 and --setMBREnable off.

Here's how I set it up:
```
it-jmccarv Release_x86_64 # ./msed --initialsetup password /dev/sda
- 16:19:53.142 INFO: takeownership complete
- 16:19:53.530 INFO: Locking SP Activate Complete
- 16:19:54.507 INFO: LockingRange0 disabled 
- 16:19:55.450 INFO: LockingRange0 set to RW
- 16:19:55.924 INFO: MBRDone set on 
- 16:19:56.393 INFO: MBREnable set on 
- 16:19:56.393 INFO: Initial setup of TPer complete on /dev/sda
it-jmccarv Release_x86_64 # ./msed --loadPBAimage password ~/msed/biospba-0.20beta.img /dev/sda
- 16:20:40.861 INFO: Writing PBA to /dev/sda
 | [*********************]
- 16:21:57.095 INFO: PBA image  /root/msed/biospba-0.20beta.img written to /dev/sda
it-jmccarv Release_x86_64 # ./msed --setMBREnable on password /dev/sda
- 16:22:11.199 INFO: MBREnable set on 
it-jmccarv Release_x86_64 # ./msed --query /dev/sda

/dev/sda ATA Samsung SSD 840 EVO 250GB                EXT0CB6Q S1DBNSAFC17239D     
TPer function (0x0001)
    ACKNAK = N, ASYNC = N. BufferManagement = N, comIDManagement  = N, Streaming = Y, SYNC = Y
Locking function (0x0002)
    Locked = N, LockingEnabled = Y, LockingSupported = Y, MBRDone = Y, MBREnabled = Y, MediaEncrypt = Y
Geometry function (0x0003)
    Align = N, Alignment Granularity = 1 (512), Logical Block size = 512, Lowest Aligned LBA = 0
Opal V1.0 function (0x0200)
Base comID = 0x1004, comIDs = 1
DataStore function (0x0202)
    Max Tables = 9, Max Size Tables = 10485760, Table size alignment = 1
OPAL 2.0 function (0x0203)
    Base comID = 0x1004, Initial PIN = 0x0, Reverted PIN = 0x0, comIDs = 1
    Locking Admins = 4, Locking Users = 9, Range Crossing = N

TPer Properties: 
  MaxComPacketSize = 66048  MaxResponseComPacketSize = 66048
  MaxPacketSize = 66028  MaxIndTokenSize = 65992  MaxPackets = 1
  MaxSubpackets = 1  MaxMethods = 1  MaxAuthentications = 5
  MaxSessions = 1  MaxTransactionLimit = 1  DefSessionTimeout = 0

Host Properties: 
  MaxComPacketSize = 2048  MaxResponseComPacketSize = 2048
  MaxPacketSize = 2028  MaxIndTokenSize = 1992  MaxPackets = 1
  MaxSubpackets = 1  MaxMethods = 1

it-jmccarv Release_x86_64 # lsblk
NAME                 MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                    8:0    0 232.9G  0 disk 
|-sda1                 8:1    0   256M  0 part 
`-sda2                 8:2    0 232.7G  0 part 
  |-vg_sys-ccache    253:0    0   2.5G  0 lvm  /var/ccache
  |-vg_sys-distfiles 253:1    0     3G  0 lvm  /usr/portage/distfiles
  |-vg_sys-home_copy 253:2    0   100G  0 lvm  /home
  |-vg_sys-opt       253:3    0    11G  0 lvm  /opt
  |-vg_sys-site      253:4    0     2G  0 lvm  /site
  |-vg_sys-src       253:5    0     3G  0 lvm  /usr/src
  |-vg_sys-tmp       253:6    0     2G  0 lvm  /tmp
  |-vg_sys-usr       253:7    0     7G  0 lvm  /usr
  |-vg_sys-var       253:8    0     3G  0 lvm  /var
  |-vg_sys-vartmp    253:9    0     8G  0 lvm  /var/tmp
  |-vg_sys-swap      253:10   0     8G  0 lvm  
  `-vg_sys-root      253:11   0     3G  0 lvm  /
sr0                   11:0    1  1024M  0 rom
```

![20150216_162708_resized](https://cloud.githubusercontent.com/assets/11036516/6220363/0fe0cf8a-b5fb-11e4-9a54-37d1edde722b.jpg)

**Comment by r0m30** [comment]
The only thing you were doing wrong was following the directions  :-(

I changed .21beta to not enable locking in the belief that it was "safer".
after you load the PBA image and enable the MBR you need to enable the
global locking range.
./msed --enableLockingRange 0 <password> <drive>


----------
Mike

**Comment by jmccarv** [comment]
Funny, I thought I remembered trying the enableLockingRange thing but apparently I remembered wrong!  That was all it took, it's working like a charm now.

I see you updated the instructions on your site too.

So was it because locking was disabled but the MBR was enabled?  I see now in the picture I took that the PBA boot shows the drive to be unlocked.  Does that mean it should be able to flip the MBR and boot without a password?

Thanks for the (very) quick response, it's appreciated!

Jason

**Comment by r0m30** [comment]
Yep, funny I thought I had put it in the directions too.

Yes, the PBA doesn't try to unlock an already unlocked drive, but the
unlock routine is also what sets MBRDone so the chain-load just reloads the
PBA.

I looked at the default ACLs in the spec and they say no you can't change
MBRdone without admin authority.  I also tried to set it as an
unauthenticated user and it failed as "NOT AUTHORIZED" which indicated that
no you can't access the data without a password because the drive will only
present the PBA until MBRDone is set.

----------
Mike

**Comment by lellis1936** [comment]
Old thread, but I had a similar issue.

I had a two-command script exit after running only the first command of these two:

msed --disablelockingrange 0
msed --setMBREnable off

The machine subsequently was rebooted, but Windows would not come up because the chain loader just reloaded the PBA, as shown in the screen shot above.

I wonder if anything can be done by PBA to make this situation a little less troublesome.  In this outcome, user must have a rescue disk and also know what to do to get an operational system again!

**Comment by r0m30** [comment]
Are you using msed or sedutil?  The newer code (Drive-Trust-Alliance/sedutil)should unshadow the MBR even if the drive is not locked.  If you are then please report it as a bug in the sedutil repo.

**Comment by lellis1936** [comment]
I *am* using sedutil, but isn't the PBA code what is at issue?  For that, I am using biospba-0.23beta.

**Comment by lellis1936** [comment]
UPDATE:  I picked up version 1.10 biospba, and tested this issue.  It indeed prevents the problem I described (Windows will not boot).

That's great, but I noticed (unless I am missing something) that neither the PBA nor --query indicates that the drive is not secure (global locking range is disabled).

The PBA continues to load, prompt for PWD report that the drive is locked as it ordinarily would.  So one would not know an enableLockingRange command should be performed to regain security on the drive.  Might it be possible/desirable to issue a warning in the boot sequence if this is the case?
